### ISSUE TITLE
FastLED v3.6.0 not sending complete LED data on ESP8266 with WiFi enabled

### ISSUE THREAD

**Issue by achronoxx** [FastLED/FastLED#1612]
I've included a simple Sketch to demonstrate a problem with certain versions of FastLED while WiFi is enabled
    - this is running on an ESP8266_NODEMCU_ESP12E
    - I observed that v3.6.0 of FastLED does not send the complete set of data to the LEDs when WiFi is connected
    - eg. this sketch controls 150 LEDS, but  FastLED is only sending 1 or 2 bytes of data, but should be sending 450 bytes (150LEDs * 3 bytes/LED)
    - FastLED.Show() is called every ~100ms, and data is sent every time,  but mostly it is 1 or 2 bytes - occasionally all 450 bytes are sent
    - I've confirmed this by looking at the data signal on a digital oscilloscope
    - it is also very evident when looking at the LED strip, as the pattern "stalls" 
    
    The problem goes away in the following 2 cases:
      1. Disable the WiFi (set wifiGUI = 0 in sketch below to test)
      2. Use v3.3.3 of FastLED (test by building with this older version)
          - I'm not sure if other versions work, as I was using v3.3.3 before I updated straight to 3.6.0

    These are the versions of all the Libraries I used:
    - Using library FastLED at version 3.6.0
    - Using library ESP8266WiFi at version 1.0 
    - Using library Hash at version 1.0 
    - Using library ESP8266mDNS at version 1.2
    - Using library ArduinoOTA at version 1.0
    - Using library NTPClient at version 3.2.0

Is there something NEW in v3.6.0 that I need to adjust to make this work, or is this a bug?

Any help would be appreciated!

Andrew

Here is the sketch that demonstrates the problem:

```
/*************************************************************************************************************************************************************
   Mar 20/24
 
    This is a simple Sketch to demonstrate problem with certain versions of FastLED while WiFi is enabled
    - this is running on an ESP8266_NODEMCU_ESP12E
    - I observed that v3.6.0 of FastLED does not send the complete set of data to the LEDs when WiFi is connected
    - eg. this sketch controls 150 LEDS, but  FastLED is only sending 1 or 2 bytes of data, but should be sending 450 bytes (150LEDs * 3 bytes/LED)
    - FastLED.Show() is called every ~100ms, and data is sent every time,  but mostly its 1 or 2 bytes - occasionally all 450 bytes are sent
    - I've confirmed this by looking at the data signal on a digital oscilloscope
    - it is also very evident when looking at the LED strip, as the pattern "stalls" 
    
    The problem goes away in the following 2 cases:
      1. Disable the WiFi (set wifiGUI = 0 in sketch below to test)
      2. Use v3.3.3 of FastLED (test by building with this older version)
          - I'm not sure if other versions work, as I was using v3.3.3 before I updated straight to 3.6.0

    These are the versions of all the Libraries I used:
    - Using library FastLED at version 3.6.0
    - Using library ESP8266WiFi at version 1.0 
    - Using library Hash at version 1.0 
    - Using library ESP8266mDNS at version 1.2
    - Using library ArduinoOTA at version 1.0
    - Using library NTPClient at version 3.2.0

**************************************************************************************************************************************************************/



char swVer[] = "24.03.20";

#define wifiGUI 0           // to enable WiFI

#if wifiGUI
#define OTA_ENABLE 1
#else
#define OTA_ENABLE 0        // make sure OTA is disabled if WiFi off
#endif

#define logger (&Serial)
#define ledESP D4
#define NUM_LEDS 150

#if wifiGUI
#define serialSel serverClient
#else
#define serialSel Serial
#endif

/*****************  NEEDED TO MAKE ESP12 NODEMCU WORK (ESP8266) *****************************************/
#define FASTLED_ESP8266_RAW_PIN_ORDER
#define FASTLED_INTERRUPT_RETRY_COUNT 0
#define FASTLED_ALLOW_INTERRUPTS 0
//#define FASTLED_ESP8266_NODEMCU_PIN_ORDER // this one does NOT work, even though I have NodeMCU board
/********************************************************************************************************/



/* =================== IMPORTANT NOTE on FastLED Version =================================== */
#include <FastLED.h>  // works with v3.3.3, but v3.6.0 causes FastLED to not send all pixels when FastLED.show() is called.  This does not happen everytime show() is called, but it happens
//                    // often and is visible in the display, as it staggers.  Note the problem goes away if WiFi is disabled.
#include <ESP8266WiFi.h>

#if wifiGUI  /////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
#include <ESP8266WiFiMulti.h>
#include <Hash.h>
ESP8266WiFiMulti WiFiMulti;

/* SSID is defined above for each individual controller, as they may be closer to different APs */
#ifndef STASSID
#define STASSID "YOUR_SSID"        // Network SSID
#define STAPSK  "YOUR_PASSWORD"                   // Network Password
#endif

//Below is for OTA programming
#if OTA_ENABLE
#include <ESP8266mDNS.h>
#include <WiFiUdp.h>
#include <ArduinoOTA.h>
#endif

//const char* ssid = STASSID;  // don't need this - use macro directly
const char* password = STAPSK;
const int port = 23;            // port to Telnet SERVER
WiFiServer server(port);
WiFiClient serverClient;

//Below is for getting Network  Time ----------------------------------------------------------------------------------------------------------------------------------------
#include <NTPClient.h>
#define utcOffsetInSeconds -18000L                                              // For UTC -5.00 : -5 * 60 * 60 = -18000 >> Toronto Time
char daysOfTheWeek[7][4] = {"Sun", "Mon", "Tue", "Wed", "Thu", "Fri", "Sat"};   // short form
//char daysOfTheWeek[7][12] = {"Sunday", "Monday", "Tuesday", "Wednesday", "Thursday", "Friday", "Saturday"};  // long form

WiFiUDP ntpUDP;       // Define NTP Client to get time

/* The pool.ntp.org automatically picks time servers which are geographically close for you. But if you want to choose explicitly, use one of the sub-zones of pool.ntp.org
    - North America:   north-america.pool.ntp.org                 */
NTPClient timeClient(ntpUDP, "north-america.pool.ntp.org", utcOffsetInSeconds);
// --------------------------------------------------------------------------------------------------------------------------------------------------------------------------

#endif // wifiGUI 


int start_led = 1;
int end_led = NUM_LEDS;
int16_t patt_dly = 95;
uint8_t glitter = 255;
uint32_t dly_deltaTime, dly_lastTime;     // used for inDelay()
char ntpTime[18];                         // day/time retrieved from NTP server


#define DATA_PIN D5       // do NOT use D0, as its connected to onboard LED on ESP12, and this drops the voltage too much for the signal to be recognized by WS2812 LEDs
#define LED_TYPE WS2812B
#define COLOR_ORDER GRB   // for strip LEDs
CRGB leds[NUM_LEDS];      // Define the array of leds


void setup() {  //ssss
  Serial.begin(115200);
  Serial.setTimeout(50);

#define BRIGHTNESS 255

  FastLED.addLeds<LED_TYPE, DATA_PIN, COLOR_ORDER>(leds, NUM_LEDS).setCorrection(TypicalLEDStrip);
  FastLED.setBrightness(BRIGHTNESS);    // set master brightness control
  FastLED.setCorrection(0xffffff);      // 0xB0B0F0
  FastLED.setDither(0);

#if wifiGUI
  setupWifi();
#endif

  pinMode(ledESP, OUTPUT); // Set up LED on ESP8266 board
} // setup


/* WiFI Setup Below =============================================================================================================================================================== */
#if wifiGUI  //   wwww===============================================================================================================================================================
void setupWifi() {

  // ====== Get ESP Version and Serial Parameters
  logger->begin(115200);
  //logger->println("\n\nUsing Serial1 for logging");
  logger->println(ESP.getFullVersion());

  // ===== Connect to WiFi
  WiFi.persistent(true);

#ifdef STASSID
  WiFiMulti.addAP(STASSID, password);  // use this if you want to add multiple SSIDs  ###### NOTE ##### Need to use WiFiMulti.run()along with this
#endif
#ifdef STASSID2
  WiFiMulti.addAP(STASSID2, password);
#endif
#ifdef STASSID3
  WiFiMulti.addAP(STASSID3, password);
#endif
#ifdef STASSID4
  WiFiMulti.addAP(STASSID4, password);
#endif
  logger->print("\nChecking for best SSID.... ");

  WiFiMulti.run();                                // use this if you want to add multiple SSIDs
  while (WiFi.status() != WL_CONNECTED) {
    logger->print('.');
    digitalWrite(ledESP, !digitalRead(ledESP));   // set built-in LED to opposite state to make it blink
    delay(500);
  }
  digitalWrite(ledESP, 0);                        // set built-in LED on if WiFi connected (0 = on)
  logger->println();
  logger->print("connected, address=");
  logger->println(WiFi.localIP());
  logger->print("SSID=");
  logger->println(WiFi.SSID());

  WiFi.mode(WIFI_STA);

  // start Telnet server
  server.begin();
  server.setNoDelay(true);
  logger->print("Ready! Use 'telnet ");
  logger->print(WiFi.localIP());
  logger->printf(" %d' to connect\n", port);


  // =======  Routine to setup OTA
#if OTA_ENABLE
  setupOTA();
#endif

  // =======  Start NTP Client to get Network Time
  timeClient.begin();
  timeClient.update();
#define DST 1                               // set to 1 to add one hour to Standard time
  uint8_t hr = timeClient.getHours();       // hours given in 24hr format, but I convert to 12hr clock below
  if (DST) {                                // add 1 hr when DST is set
    if (hr < 23) hr++;                      // check for wraparound
    else hr = 0;
  }
  sprintf(ntpTime, "%s, %02d:%02d:%02d %s", daysOfTheWeek[timeClient.getDay()], hr > 12 ? (hr - 12) : hr, timeClient.getMinutes(), timeClient.getSeconds(), hr >= 12 ? "pm" : "am");
  serialSel.printf("Start Time: %s\n", ntpTime);

} // end setupWifi()



void setupOTA() {
  // Port defaults to 8266
  // ArduinoOTA.setPort(8266);

  // Hostname defaults to esp8266-[ChipID]
  // ArduinoOTA.setHostname("myesp8266");

  // No authentication by default
  // ArduinoOTA.setPassword((const char *)"123");

  ArduinoOTA.onStart([]() {
    Serial.println("Start");
  });
  ArduinoOTA.onEnd([]() {
    Serial.println("\nEnd");
  });
  ArduinoOTA.onProgress([](unsigned int progress, unsigned int total) {
    Serial.printf("Progress: %u%%\r", (progress / (total / 100)));
  });
  ArduinoOTA.onError([](ota_error_t error) {
    Serial.printf("Error[%u]: ", error);
    if (error == OTA_AUTH_ERROR) Serial.println("Auth Failed");
    else if (error == OTA_BEGIN_ERROR) Serial.println("Begin Failed");
    else if (error == OTA_CONNECT_ERROR) Serial.println("Connect Failed");
    else if (error == OTA_RECEIVE_ERROR) Serial.println("Receive Failed");
    else if (error == OTA_END_ERROR) Serial.println("End Failed");
  });
  ArduinoOTA.begin();
  Serial.println("Ready");
  Serial.print("IP address: ");
  Serial.println(WiFi.localIP());
}//end setupOTA()

#endif // wifiGUI



void loop() {
  // Colour Glitter - It turns off previous pattern by clearing strip, and then displays 3 colours of glitter
  inDelay(patt_dly);              /* function that reads serial input and checks for WiFi clients while processing delay */
  FastLED.clearData();            // clear all LEDs so only glitter appears
  ColourGlitter(glitter);
  FastLED.show();

}


void keyIn() {
  if (serialSel.available() > 0) {
    char key = serialSel.read();
    if (key == 'R') printWiFiGUI();                 // for convenience, the parameter values will be printed if rndPattEnable is selected
  }
} // end keyIn


void printWiFiGUI() {

  serialSel.printf("SW VERSION: %s\n", swVer);

#if wifiGUI
  serialSel.printf("WiFi Status: %d (3 = Connected) RSSI: %d dBm\nIP=", WiFi.status(), WiFi.RSSI());
  serialSel.println(WiFi.localIP());
  serialSel.print("SSID: ");
  serialSel.println(WiFi.SSID());
#endif
  serialSel.printf("Start Time: %s   \n", ntpTime);
  serialSel.printf("Free heap: %d\n", ESP.getFreeHeap());

} // printWiFiGUI



void ColourGlitter(fract8 chanceOfGlitter)
{
  if (random8() < chanceOfGlitter)
  {
    leds[random16(end_led - start_led) + start_led - 1] += CRGB::White;
    leds[random16(end_led - start_led) + start_led - 1] += CRGB::Red;
    leds[random16(end_led - start_led) + start_led - 1] += CRGB::Green;
  }
}



void inDelay(int16_t dly_ms) { /*  This is a delay function that reads the Serial Port for User Input  */

  dly_lastTime = millis();

  ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
  do {                                        // do-loop used to enter at least once to read serial data
    dly_deltaTime = millis() - dly_lastTime;

    ESP.wdtFeed();                            // update ESP SW watch dog timer (WDT) (seems to be 3sec window)

#if wifiGUI
    newClient();                              // check for client connection to SERVER
#endif

#if OTA_ENABLE
    ArduinoOTA.handle();                      // check for OTA programming
#endif

    keyIn();                                  // read/process keyboard or BT Serial data

  } while (dly_deltaTime < dly_ms);
  ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////

} // inDelay()



#if wifiGUI
void newClient() {                  // function to check for new client connection
  delay(1);                         // allows WiFi to run for watchdog; also found delay had to be at least 1ms to reliably detect new client

  // ====== check if there are any new clients
  if (server.hasClient()) {
    serverClient.stop();            // disconnect current client, as sometimes connection is lost, and reconnect does not work as it thinks it's still connected

    // find free/disconnected spot
    if (!serverClient.connected()) {
      serverClient = server.available();
      logger->println(F("\nNew client connected...\n"));
    }

    // no free/disconnected spot so reject
    else {
      server.available().println("busy");
      // hints: server.available() is a WiFiClient with short-term scope
      // when out of scope, a WiFiClient will
      // - flush() - all data will be sent
      // - stop() - automatically too
      logger->printf("server is busy with active connection...\n");
    }
  } // if(server.hasClient)

} // end newClient()


#endif // for WiFi ==============================================================================================================



```

**Comment by zackees** [comment]
Hi Andrew, this is an issue with the 4.4 RMT driver. Unfortunately this is a legacy driver that is going away, and we have no plans on touching this code much more than we already have as it has been finely tuned to work the way it has.

It's entirely possible that the issues you are describing have also been addressed in an update to FastLED, since the version you are referencing is quite old.

**Comment by achronoxx** [comment]
Hi Zach,

Thanks for the response.  I will try the latest version in the next few days and report back.

Andrew

**Comment by zackees** [comment]
RMT5 has been released on master.

You'll need idf5.1 or higher.

**Comment by zackees** [comment]
Closing this big as i think it's a solved issue with the release of 3.9.0
